libvast/include/vast/detail/inspection_common.hpp

```diff
@@ -0,0 +1,109 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include <string_view>
+#include <utility>
+
+namespace vast::detail {
+
+template <class Inspector, class AfterInspectionCallback>
+class inspection_object_with_after_inspection_callback;
+
+/// Common class used by VAST inspectors as a return type for object method.
+/// The CAF inspector API requires an inspector to have object method.
+/// This method should return an object which guides the inspection procedure.
+/// Example of usage:
+/// bool inspect(Inspector& f, some_type& x)
+///{
+///   auto cb = []{};
+///   return f.object(x).
+///     on_load(cb).
+///     fields(f.field("name", x.name), f.field("value", x.value));
+///}
+/// Such code should result in inspection of x.name by the inspector. If the
+/// inspection succeeds then it should proceed to inspection of x.value. If all
+/// provided fields succeed then callback (cb variable) should be called. The
+/// provided string literals are optionally used by human readable CAF
+/// inspectors. Current VAST inspectors have no usage for them.
+template <class Inspector>
+class inspection_object {
+public:
+  explicit inspection_object(Inspector& inspector) : inspector_{inspector} {
+  }
+
+  template <class... Fields>
+  constexpr bool fields(Fields&&... fields) {
+    return (fields(inspector_) && ...);
+  }
+
+  constexpr inspection_object& pretty_name(std::string_view) noexcept {
+    return *this;
+  }
+
+  template <class Callback>
+    requires(Inspector::is_loading == true)
+  constexpr decltype(auto) on_load(Callback callback) {
+    return inspection_object_with_after_inspection_callback{
+      inspector_, std::move(callback)};
+  }
+
+  template <class Callback>
+    requires(Inspector::is_loading == false)
+  constexpr decltype(auto) on_save(Callback callback) {
+    return inspection_object_with_after_inspection_callback{
+      inspector_, std::move(callback)};
+  }
+
+private:
+  Inspector& inspector_;
+};
+
+/// Enhanced inspection_object that is responsible for calling callback after
+/// inspection.
+template <class Inspector, class AfterInspectionCallback>
+class inspection_object_with_after_inspection_callback {
```

Comment:
This class name is quite the mouth-full. Can we unify the two and default initialize `AfterInspectionCallback` to a noop?

Reply:
It started with being unified. I just thought that maybe a performance overhead for extra unused dummy member might be unwanted. In the end the compiler can optimize it and my worry is premature optimization. I could just maybe rename it to inspection_object_with_callback ?

Reply:
> In the end the compiler might optimize it and my worry is premature optimization.

I would assume that. As long as the no-op's definition is visible it should be extremely easy for the compiler to skip the call.

Reply:
Done